File: leetcode/025_reverse_nodes_in_k_group.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func reversePart(head *ListNode, end *ListNode) *ListNode {
	var p *ListNode
	c := head
	n := c
	for c != end {
		n = n.Next
		c.Next = p
		p = c
		c = n
	}
	return p
}

func reverseKGroup(head *ListNode, k int) *ListNode {
	if k <= 1 {
		return head
	}
	var p *ListNode
	c := head
	for {
		begin := c
		end := c
		for i := 0; i < k; i++ {
			if end == nil {
				return head
			}
			end = end.Next
		}
		h := reversePart(begin, end)
		if p == nil {
			head = h
		} else {
			p.Next = h
		}
		p = begin
		p.Next = end
		c = end
	}
}

func makeListNode(a []int) *ListNode {
	if len(a) == 0 {
		return nil
	}
	ret := &ListNode{}
	c := ret
	for _, i := range a {
		c.Next = &ListNode{
			Val: i,
		}
		c = c.Next
	}
	return ret.Next
}

func getListNode(head *ListNode) []int {
	var ret []int
	for head != nil {
		ret = append(ret, head.Val)
		head = head.Next
	}
	return ret
}

func main() {
	tests := []struct {
		input  []int
		k      int
		output []int
	}{
		{nil, 1, nil},
		{nil, 2, nil},
		{[]int{1, 2, 3, 4, 5}, 1, []int{1, 2, 3, 4, 5}},
		{[]int{1, 2, 3, 4, 5}, 2, []int{2, 1, 4, 3, 5}},
		{[]int{1, 2, 3, 4, 5}, 3, []int{3, 2, 1, 4, 5}},
		{[]int{1, 2, 3, 4, 5}, 4, []int{4, 3, 2, 1, 5}},
		{[]int{1, 2, 3, 4, 5}, 5, []int{5, 4, 3, 2, 1}},
	}
	for _, test := range tests {
		list := makeListNode(test.input)
		ret := reverseKGroup(list, test.k)
		a := getListNode(ret)
		fmt.Println(a, test.output)
	}
}

File: leetcode/025_reverse_nodes_in_k_group.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if k < 2:
            return head
        p = None
        c = head
        while True:
            print getList(head)
            begin = c
            end = c
            i = 0
            while end!=None and i<k:
                i += 1
                end = end.next
            if i != k:
                if p != None:
                    p.next = begin
                return head
            newHead = self.reverse(begin, end)
            if p == None:
                head = newHead
            else:
                p.next = newHead
            p = begin
            c = end

    def reverse(self, head, end):
        p = None
        c = head
        while c != end:
            n = c.next
            c.next = p
            p = c
            c = n
        return p

def mkList(a):
    if len(a) == 0:
        return None
    ret = ListNode(0)
    c = ret
    for i in a:
        c.next = ListNode(i)
        c = c.next
    return ret.next

def getList(head):
    ret = []
    while head != None:
        ret += [head.val]
        head = head.next
    return ret

l = mkList([1,2,3,4,5])
s = Solution()
h = s.reverseKGroup(l, 6)
a = getList(h)
print a
